File: csharp/CsSearch/CsSearch/FileUtil.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace CsSearch
{
	public static class FileUtil
	{
		private const string CurrentPath = ".";
		private const string ParentPath = "..";
		private static readonly ISet<string> DotDirs = new HashSet<string> { CurrentPath, ParentPath };

		private const char ForwardSlash = '/';
		private const char BackSlash = '\\';
		private static readonly char[] DirSeps = new char[] { ForwardSlash, BackSlash };

		public static string GetHomePath()
		{
			return Environment.GetEnvironmentVariable("HOME")
			       ?? Environment.GetEnvironmentVariable("USERPROFILE")
			       ?? "~";
		}

		public static IEnumerable<string> EnumerableStringFromFile(SearchFile f, Encoding enc)
		{
			return EnumerableStringFromFile(f.FullName, enc);
		}

		public static IEnumerable<string> EnumerableStringFromFile(string filepath, Encoding enc)
		{
			using var sr = new StreamReader(filepath, enc);
			// read each line, ensuring not null (EOF)
			string? line;
			while ((line = sr.ReadLine()) != null)
			{
				// return trimmed line
				yield return line;
			}
		}

		public static string GetFileContents(string filepath, Encoding encoding)
		{
			try
			{
				using var sr = new StreamReader(filepath, encoding);
				var contents = sr.ReadToEnd();
				return contents;
			}
			catch (IOException e)
			{
				throw new SearchException(e.Message);
			}
		}

		public static string GetFileContents(SearchFile f, Encoding encoding)
		{
			return GetFileContents(f.FullName, encoding);
		}

		public static string ExpandPath(string filepath)
		{
			return filepath[0] == '~' ? JoinPath(GetHomePath(), filepath.Substring(1)) : filepath;
		}

		public static string ContractPath(string filepath)
		{
			return filepath[0] == '~' ? filepath : filepath.Replace(GetHomePath(), "~");
		}

		public static string GetRelativePath(string fullPath, string startpath)
		{
			var filePath = NormalizePath(fullPath);
			startpath = NormalizePath(startpath);
			var startFullPath = NormalizePath(new DirectoryInfo(startpath).FullName);
			if (startFullPath != startpath)
			{
				filePath = filePath.Replace(startFullPath, startpath);
			}
			return filePath;
		}

		public static string ContractOrRelativePath(string fullPath, string startpath)
		{
			if (startpath[0] == '~')
			{
				return ContractPath(fullPath);
			}
			return GetRelativePath(fullPath, startpath);
		}

		public static bool IsDotDir(string filename)
		{
			return DotDirs.Contains(NormalizePath(filename));
		}

		public static bool IsHidden(FileSystemInfo f)
		{
			return ((f.Name.StartsWith(CurrentPath) && !IsDotDir(f.Name))
			        || (f.Exists && (f.Attributes & FileAttributes.Hidden) != 0));
		}

		public static string JoinPath(string path1, string path2)
		{
			var dirSep = ForwardSlash;
			if (path1.IndexOf(BackSlash) > -1)
				dirSep = BackSlash;
			if (path2[0] == ForwardSlash || path2[0] == BackSlash)
				path2 = path2.Substring(1);
			return NormalizePath(path1) + dirSep + path2;
		}

		public static string NormalizePath(string path)
		{
			return path.TrimEnd(DirSeps);
		}
	}
}

File: fsharp/FsSearch/FsSearchLib/FileUtil.fs
﻿namespace FsSearchLib

open System
open System.IO
open System.Text
open System.Text.RegularExpressions

module FileUtil = 

    let currentPath = "."
    let parentPath = ".."
    let forwardSlash = '/'
    let backSlash = '\\'
    let dirSeps = [| forwardSlash; backSlash |]
    let dotDirs = Set.ofList [
        currentPath; parentPath;
        currentPath + string forwardSlash; parentPath + string forwardSlash;
        currentPath + string backSlash; parentPath + string backSlash
    ]

    let GetHomePath () : string = 
        match Environment.GetEnvironmentVariable("HOME") with
        | home when home <> null -> home
        | _ -> Environment.GetEnvironmentVariable("USERPROFILE")

    let GetFileContents (filepath : string) (encoding : Encoding) : string =
        let contents =
            try
                use sr = new StreamReader (filepath, encoding)
                sr.ReadToEnd()
            with
            | :? IOException as e -> printfn $"%s{e.Message}"; ""
        contents

    let GetFileLines (filePath : string) (encoding : Encoding) = seq {
        use sr = new StreamReader (filePath, encoding)
        while not sr.EndOfStream do
            yield sr.ReadLine ()
    }

    let NormalizePath (path : string) : string = 
        path.TrimEnd(dirSeps)

    let JoinPath (path1 : string) (path2 : string) : string = 
        let dirSep =
            if path1.IndexOf(backSlash) > -1 then backSlash else forwardSlash
        let p2 =
            if path2.[0] = forwardSlash || path2.[0] = backSlash then path2.Substring(1) else path2
        String.Format("{0}{1}{2}", NormalizePath path1, dirSep, p2)

    let ExpandPath (filepath : string) : string =
        if filepath.[0] = '~' then JoinPath (GetHomePath()) (filepath.Substring(1))
        else filepath

    let ContractPath (filepath : string) : string =
        if filepath.[0] = '~' then filepath 
        else filepath.Replace(GetHomePath(), "~")

    let GetRelativePath (fullpath : string) (startpath : string) : string =
        let startFullPath = NormalizePath (DirectoryInfo(startpath)).FullName
        let normStartPath = NormalizePath startpath
        if startFullPath <> normStartPath
        then fullpath.Replace (startFullPath, normStartPath)
        else fullpath

    let ContractOrRelativePath (fullpath : string) (startpath : string) : string =
        if startpath.[0] = '~' then ContractPath fullpath 
        else GetRelativePath fullpath startpath
        
    let IsDotDir (filepath : string): bool = dotDirs.Contains(filepath)

    let IsHidden (filepath : string) : bool = 
        let startsWithDot = filepath.[0] = '.' && not (IsDotDir filepath)
        //let hasHiddenAttribute = f.Exists && (f.Attributes &&& FileAttributes.Hidden) <> 0
        startsWithDot

    let IsHiddenFile (f : FileSystemInfo) : bool = 
        (f.Name.[0] = '.' && not (IsDotDir f.Name)) ||
        (f.Exists && (f.Attributes &&& FileAttributes.Hidden) = FileAttributes.Hidden)

    let ExtensionsListFromString (exts : string) : string list =
        let nonWord = Regex(@"\W+")
        nonWord.Split(exts)
        |> Array.toList
        |> List.filter (fun (x : string) -> String.IsNullOrEmpty(x) = false)
        |> List.map (fun (x : string) -> if (x.StartsWith(".")) then x else "." + x)

;;
